inspector/src/components/graph/canvasGraphService.ts

```diff
@@ -416,9 +459,131 @@ export class CanvasGraphService {
     }
+    
+    /**
+     * Handles what to do when we are hovering over the canvas and not panning.
+     * 
+     * @param event A reference to the event to be handled. 
+     */
+    private _handleDataHover = (event: MouseEvent) => {
+        if (this._panPosition) {
+            // we don't want to do anything if we are in the middle of panning
+            return;
+        }
+
+        this._hoverPosition = event.clientX;
+
+        // then draw the tooltip.
+        this._drawTooltip(this._hoverPosition, this._drawableArea);
+    }
+
+    /**
+     * Handles what to do when we stop hovering over the canvas.
+     */
+    private _handleStopHover = () => {
+        this._hoverPosition = null;
+    }
+
+    /**
+     * Draws the tooltip given the area it is allowed to draw in and the current pixel position.
+     * 
+     * @param pixel the position of the mouse cursor in pixels. 
+     * @param drawableArea  the available area we can draw in.
+     */
+    private _drawTooltip(pixel: number | null, drawableArea: IGraphDrawableArea) {
+        const {_ctx : ctx} = this;
+
+        if (pixel === null || !ctx || !ctx.canvas) {            
+            return;
+        }
+
+        // first convert the mouse position in pixels to a timestamp.
+        const {left: start, right: end} = ctx.canvas.getBoundingClientRect();
+        const inferredTimestamp = this._getNumberFromPixel(pixel, this._globalTimeMinMax, start, end);
+        
+        const results: IPerfTooltip[] = [];
+
+        // get the closest timestamps to the target timestamp, and store the appropriate meta object.
+        this.datasets.forEach((dataset: IPerfDataset) => {
+            if (!!dataset.hidden || dataset.data.length === 0) {
+                return;
+            }
+
+            const closestIndex = this._getClosestPointToTimestamp(dataset, inferredTimestamp);
+            
+            results.push({
+                            text: `${dataset.id}: ${dataset.data[closestIndex].value.toFixed(2)}`,
+                            color: dataset.color ?? defaultColor,
+                        });
+        });
+
+        const longestText = results.reduce((longestSoFar: string, result) => {
+            if (longestSoFar.length < result.text.length) {
+                return result.text;
+            } else {
+                return longestSoFar;
+            }
+        }, "");
+
+        let x = pixel - start;
+        let y = Math.floor((drawableArea.bottom - drawableArea.top)/2);
+
+        ctx.save();
+        
+        ctx.font = tooltipFont;
+        ctx.textBaseline = "middle";
+        ctx.textAlign = "left";
+
+        const boxLength = this._tooltipLineHeight;
+        const textHeight = this._tooltipLineHeight + Math.floor(tooltipHorizontalPadding/2);
+
+        const width = ctx.measureText(longestText).width + boxLength + 2 * tooltipHorizontalPadding + spaceBetweenTextAndBox;

```

Comment:
I wonder how measureText speed is but depending on your findings, might be good to cache ?

Reply:
So the performance of measureText depends entirely on the length of the text: https://jsbench.me/d4kr47hmda/1

If you look at the "Small" section it seems to be very very fast (for me it is 0.01ms or 72126.07 ops/s), and around the length of what our text would be at largest for our predefined tags. 

Although the performance does degrade pretty hard for the "large" section as compared to the small, because of this I could do some form of caching just in case, if we follow the no gain is too small rule!

Reply:
Latest commit caches this!